File: sdks/python/durable/contract.py
"""Agent contracts — enforceable invariants at the step boundary."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable


@dataclass
class Contract:
    """A named contract that validates tool calls before execution.

    The check function receives ``(step_name, args)`` and should raise
    ``ValueError`` if the invariant is violated. The agent suspends for
    human review — the tool does not execute.

    Example::

        Contract("max-charge", lambda step, args: (
            None if args.get("amount", 0) <= 100
            else (_ for _ in ()).throw(ValueError("too much"))
        ))

    Or more readably, use the ``@agent.contract`` decorator.
    """

    name: str
    check: Callable[[str, dict], None]

File: src/agent/contract.rs
//! Agent contracts — enforceable invariants at the step boundary.
//!
//! Contracts are checked AFTER the LLM decides to call a tool but BEFORE
//! the tool executes. A violation triggers suspension (not crash), allowing
//! human review. Contract checks are recorded as events for auditability.

use crate::json::Value;
use std::sync::Arc;

/// A named contract that validates tool calls before execution.
///
/// The check function receives `(step_name, step_args)` and returns
/// `Ok(())` if the invariant holds, or `Err(reason)` if violated.
#[derive(Clone)]
pub struct Contract {
    /// Human-readable name for this contract (e.g., "max-charge").
    pub name: String,
    /// The validation function.
    pub check: Arc<dyn Fn(&str, &Value) -> Result<(), String> + Send + Sync>,
}
